### PR TITLE
Add new subroutine and filter out disable networks for switchdiscover

### PR DESCRIPTION
**Pull request by cxhong**
This PR covers issue #1106 and issue #1038 
1) Process osscan-guess to find vendor info  if vendor info showed as "IBM" (or other than global_switch_type defined in the code)
2) created new subroutine nmap_osguess() so we can run nmap osscan-guess for multiple ips instead of one at a time.
3) only run switchdiscover for not disable networks if there is no input from --range options.

### CODE REVIEW

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm [hunk 1]
```diff
@@ -716,77 +719,115 @@ sub nmap_scan {
...
+                    if ( ( $line =~ /switch/) || ( $line =~ /router/) ) {
+                        $isswitch=1;
+                    } else {
+                        last;
+                    }  
+                }
+                my $search_string = join '|', keys(%global_switch_type);
+                if ($line =~ /Running/) {
+                    if ($line =~ /($search_string)/){
+                        $vendor = $1;
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: Do we need to set ``$isswitch=1`` here?

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm [hunk 2]
```diff
@@ -716,77 +719,115 @@ sub nmap_scan {
...
+                if ($line =~ /Aggressive OS/) {
+                    if ($line =~ /($search_string)/){
+                        $vendor = $1;
+                        $isswitch=1;
+                    }
+                }
+                
+            }
+            if ($isswitch == 1) {
+                $switches->{$mac}->{ip} = $ip;
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: what if ``ip`` or ``vendor`` is blank?

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm [hunk 3]
```diff
@@ -1164,13 +1202,42 @@ sub get_ip_ranges {
...
+    # my_nets will return all the networks defined in the networks 
+    # table, we need to exclude disable networks here.
     my $nets = xCAT::NetworkUtils->my_nets();
     my $ranges=[];
+    my $nettab = xCAT::Table->new('networks');
+    my $disable_nets=[];
+    if ($nettab) {
+        my $netents = $nettab->getAllEntries("all");
+        foreach (@$netents) {
+            if ($_->{disable} =~ /1|YES|Yes|yes|Y|n/) {
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: why check the ``|n/``  in line ``if ($_->{disable} =~ /1|YES|Yes|yes|Y|n/) {``? ``n`` means NO.

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm [hunk 4]
```diff
@@ -1164,13 +1202,42 @@ sub get_ip_ranges {
...
+                push(@$disable_nets, $_->{'net'});
+                if (exists($globalopt{verbose}))    {
+                    send_msg($request, 1, "network is disabled for  $_->{'net'}, $_->{'mask'}");
+                }
+            }
+        }
+    }
```
Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: I am concerning the algorithm that exclude the 'disabled' network in the ``networks`` table.

Generally, user only put the useful network in the ``networks`` table. So I suggest that we only scan the networks that enabled in the ``networks`` table.

Could you refer to the ``lsdef`` command?

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm (hunk 1 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: somehow this output is a little different depends version of nmap, OS or switch
``````
Device type:
Running (JUST GUESSING):
Aggressive OS guesses:
`````
for above three "keywords" line, sometime has two of three, sometime has one of three, the results are inconsistent.

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm (hunk 2 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: ip address will be always here.  if vendor is blank, but Device type showed as switch, we will consider this as switch.

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm (hunk 3 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: oh, right, i will change that

File: xCAT-server/lib/xcat/plugins/switchdiscover.pm (hunk 4 above)
Comment (cxhong) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: not sure how can I check if network is disabled via other method.

Comment (daniceexi) [inline xCAT-server/lib/xcat/plugins/switchdiscover.pm]: In fact, the call ``$nettab->getAllEntries()`` without argument will only return the enabled networks.

My question is if certain MN has 10 nics, but only defined the install_network and service_network in the networks table, do we want to discover against the 10 nics or just the two networks which have been defined in the networks table? My idea is we just discover the enabled networks which defined in the networks table.

**Comment by cxhong** [comment]
@daniceexi , I agree we should get enabled networks from networks table.  Modify the code, can you review it again?

**Comment by daniceexi** [comment]
I have reviewed the code change and agree to merge. Thanks @cxhong

**Pull request merged**